File: src/policy.py
"""Policy validation for workflow transitions.

This module handles YAML frontmatter parsing and policy enforcement
for per-state transition restrictions.
"""
import re
import logging
import yaml
from pathlib import Path
from typing import Dict, Any, List, Optional, Union
from dataclasses import dataclass
from .parsing import Transition

logger = logging.getLogger(__name__)


# Supported extensions for state files
STATE_EXTENSIONS = {'.md', '.sh', '.bat'}


@dataclass
class Policy:
    """Represents a workflow state's transition policy.

    Attributes:
        allowed_transitions: List of allowed transition combinations.
            Each entry is a dict with 'tag' and optionally 'target', 'return', 'next', etc.
            Example: [{"tag": "goto", "target": "NEXT.md"}, {"tag": "result"}]
        model: Optional model specification from frontmatter (e.g., "opus", "sonnet", "haiku").
            If specified, this model will be used for Claude Code invocations for this state.
        effort: Optional effort level from frontmatter (e.g., "low", "medium", "high").
            If specified, this effort level will be used for Claude Code invocations for this state.
    """
    allowed_transitions: List[Dict[str, Any]]
    model: Optional[str] = None
    effort: Optional[str] = None


class PolicyViolationError(ValueError):
    """Raised when a transition violates the state's policy."""
    pass


def _targets_match(policy_target: str, transition_target: str) -> bool:
    """Check if a transition target matches a policy target.

    Handles abstract state names in policies. If the policy target has no
    extension, it matches any resolved form of that state name.

    Examples:
        - policy "COUNT" matches transition "COUNT.sh", "COUNT.bat", "COUNT.md"
        - policy "COUNT.md" matches only transition "COUNT.md"
        - policy "COUNT.sh" matches only transition "COUNT.sh"

    Args:
        policy_target: Target specified in policy (may be abstract or explicit)
        transition_target: Resolved target from transition (always has extension)

    Returns:
        True if the targets match, False otherwise
    """
    # Exact match always works
    if policy_target == transition_target:
        return True

    # Check if policy target is abstract (no extension)
    policy_path = Path(policy_target)
    policy_ext = policy_path.suffix.lower()

    if policy_ext:
        # Policy has explicit extension - only exact match allowed (already checked above)
        return False

    # Policy target is abstract - check if transition matches it
    # The transition target should be the policy target with a valid extension
    transition_path = Path(transition_target)
    transition_stem = transition_path.stem
    transition_ext = transition_path.suffix.lower()

    # Match if stems are equal and transition has a valid state extension
    return transition_stem == policy_target and transition_ext in STATE_EXTENSIONS


def should_use_reminder_prompt(policy: Optional[Policy]) -> bool:
    """Check if reminder prompt should be used for parse failures.
    
    Reminder prompts are only used when:
    - Policy exists
    - Policy has non-empty allowed_transitions list
    
    If no policy or no allowed_transitions, parse failures should
    terminate the agent with an error.
    
    Args:
        policy: The policy to check (None means no policy)
        
    Returns:
        True if reminder prompt should be used, False if error should be raised
    """
    if policy is None:
        return False
    
    if not policy.allowed_transitions:
        return False
    
    return True


def parse_frontmatter(content: str) -> tuple[Optional[Policy], str]:
    """Parse YAML frontmatter from markdown content.
    
    Frontmatter is expected to be at the start of the file, delimited by `---`
    on separate lines. If no frontmatter is present, returns (None, content).
    
    Args:
        content: The full file content (may include frontmatter)
        
    Returns:
        Tuple of (Policy object or None, body content without frontmatter)
        
    Raises:
        ValueError: If frontmatter is malformed YAML
    """
    # Pattern to match frontmatter: ---\n...\n---
    # Must match at the start of content
    # Handle both empty frontmatter (---\n---\n) and non-empty (---\ncontent\n---\n)
    # First try to match with content, then try empty frontmatter
    frontmatter_pattern = re.compile(r'^---\s*\n(.+?)\n---\s*\n', re.DOTALL)
    empty_frontmatter_pattern = re.compile(r'^---\s*\n---\s*\n', re.DOTALL)
    
    match = frontmatter_pattern.match(content)
    if match:
        yaml_content = match.group(1)
        body = content[match.end():]
    else:
        # Try empty frontmatter
        empty_match = empty_frontmatter_pattern.match(content)
        if empty_match:
            yaml_content = ""
            body = content[empty_match.end():]
        else:
            # No frontmatter found
            return None, content
    
    # Parse empty frontmatter as no policy
    if not yaml_content.strip():
        return None, body
    
    try:
        data = yaml.safe_load(yaml_content)
        if not data:
            return None, body
        
        # Extract allowed_transitions (default to empty list)
        allowed_transitions = data.get("allowed_transitions", [])
        if not isinstance(allowed_transitions, list):
            allowed_transitions = []
        
        # Validate that each entry is a dict with at least a 'tag' key
        validated_transitions = []
        for entry in allowed_transitions:
            if isinstance(entry, dict) and "tag" in entry:
                validated_transitions.append(entry)
        
        # Extract model field (optional)
        model = data.get("model")
        if model is not None:
            if not isinstance(model, str):
                # Model must be a string if present
                model = None
            else:
                # Normalize: strip whitespace and convert to lowercase
                model = model.strip().lower()
                # Treat empty strings as None
                if not model:
                    model = None
                # Validate model value (only for Claude Code currently)
                # Valid values: opus, sonnet, haiku
                elif model not in ("opus", "sonnet", "haiku"):
                    # Invalid model - log warning but don't fail
                    # We'll let Claude Code handle invalid models
                    # This allows for future expansion to other agent CLIs
                    logger.warning(
                        f"Unknown model '{model}' in frontmatter. "
                        f"Valid values for Claude Code: opus, sonnet, haiku. "
                        f"Passing to Claude Code as-is."
                    )

        # Extract effort field (optional)
        effort = data.get("effort")
        if effort is not None:
            if not isinstance(effort, str):
                # Effort must be a string if present
                effort = None
            else:
                # Normalize: strip whitespace and convert to lowercase
                effort = effort.strip().lower()
                # Treat empty strings as None
                if not effort:
                    effort = None
                # Validate effort value
                # Valid values: low, medium, high
                elif effort not in ("low", "medium", "high"):
                    # Invalid effort - log warning but don't fail
                    logger.warning(
                        f"Unknown effort level '{effort}' in frontmatter. "
                        f"Valid values for Claude Code: low, medium, high. "
                        f"Passing to Claude Code as-is."
                    )

        policy = Policy(allowed_transitions=validated_transitions, model=model, effort=effort)
        
        return policy, body
    except yaml.YAMLError as e:
        raise ValueError(f"Invalid YAML frontmatter: {e}") from e


def validate_transition_policy(transition: Transition, policy: Optional[Policy]) -> None:
    """Validate that a transition complies with the state's policy.
    
    Args:
        transition: The transition to validate
        policy: The policy to validate against (None means no restrictions)
        
    Raises:
        PolicyViolationError: If the transition violates the policy
    """
    # No policy means no restrictions
    if policy is None:
        return
    
    # If no allowed_transitions specified, allow all
    if not policy.allowed_transitions:
        return
    
    # Check if transition matches any allowed entry
    for allowed in policy.allowed_transitions:
        # Tag must match
        if allowed.get("tag") != transition.tag:
            continue
        
        # For result tag, no target/attributes to check
        if transition.tag == "result":
            return  # Match found
        
        # Check target matches (if specified in policy)
        # Use _targets_match to allow abstract policy targets (e.g., "COUNT")
        # to match resolved transition targets (e.g., "COUNT.sh")
        if "target" in allowed:
            if not _targets_match(allowed["target"], transition.target):
                continue

        # Check all other attributes match
        # For call/function: check return attribute (also a state reference)
        if "return" in allowed:
            transition_return = transition.attributes.get("return", "")
            if not _targets_match(allowed["return"], transition_return):
                continue

        # For fork: check next attribute (also a state reference)
        if "next" in allowed:
            transition_next = transition.attributes.get("next", "")
            if not _targets_match(allowed["next"], transition_next):
                continue
        
        # All checks passed - this is a valid transition
        return
    
    # No matching entry found
    # Build helpful error message
    allowed_for_tag = [a for a in policy.allowed_transitions if a.get("tag") == transition.tag]
    if allowed_for_tag:
        # Tag is allowed but this specific combination isn't
        raise PolicyViolationError(
            f"Transition '{transition.tag}' with target '{transition.target}' "
            f"and attributes {transition.attributes} is not allowed. "
            f"Allowed combinations for '{transition.tag}': {allowed_for_tag}"
        )
    else:
        # Tag itself is not allowed
        allowed_tags = list(set(a.get("tag") for a in policy.allowed_transitions))
        raise PolicyViolationError(
            f"Tag '{transition.tag}' is not allowed. "
            f"Allowed tags: {allowed_tags}"
        )


def can_use_implicit_transition(policy: Optional[Policy]) -> bool:
    """Check if implicit transition optimization can be used.
    
    Implicit transitions can be used when:
    - Policy exists and has exactly one allowed transition
    - The transition is NOT a result tag (result tags have variable payload)
    - All required information (tag, target, attributes) is predetermined
    
    Args:
        policy: The policy to check (None means no policy)
        
    Returns:
        True if implicit transition can be used, False otherwise
    """
    # No policy means we can't use implicit transitions
    if policy is None:
        return False
    
    # Empty policy means no restrictions, can't use implicit
    if not policy.allowed_transitions:
        return False
    
    # Must have exactly one allowed transition
    if len(policy.allowed_transitions) != 1:
        return False
    
    # Get the single allowed transition
    allowed = policy.allowed_transitions[0]
    tag = allowed.get("tag")
    
    # Result tags cannot be implicit (payload is variable)
    if tag == "result":
        return False
    
    # For non-result tags, we need at least a target
    # If target is missing, we can't construct the transition
    if "target" not in allowed:
        return False
    
    # All checks passed - can use implicit transition
    return True


def get_implicit_transition(policy: Optional[Policy]) -> Transition:
    """Get the implicit transition from a policy.
    
    This function constructs a Transition object from the single allowed
    transition in the policy. It should only be called when
    can_use_implicit_transition() returns True.
    
    Args:
        policy: The policy with exactly one non-result allowed transition
        
    Returns:
        Transition object constructed from the policy
        
    Raises:
        ValueError: If policy is None, empty, has multiple transitions,
                   or the only transition is a result tag
    """
    if not can_use_implicit_transition(policy):
        raise ValueError(
            "Cannot get implicit transition: policy must have exactly one "
            "non-result allowed transition"
        )
    
    allowed = policy.allowed_transitions[0]
    tag = allowed.get("tag")
    target = allowed.get("target", "")
    
    # Build attributes dictionary (exclude 'tag' and 'target' which are not attributes)
    attributes = {}
    for key, value in allowed.items():
        if key not in ("tag", "target"):
            attributes[key] = value
    
    return Transition(
        tag=tag,
        target=target,
        attributes=attributes,
        payload=""
    )


def generate_reminder_prompt(policy: Policy) -> str:
    """Generate a reminder prompt from allowed transitions.
    
    This function creates a helpful reminder message that lists all
    allowed transitions for the current state. It formats the transitions
    in a clear, readable way that can be appended to the original prompt.
    
    Args:
        policy: The policy containing allowed_transitions
        
    Returns:
        A formatted reminder string listing all allowed transitions
        
    Raises:
        ValueError: If policy is None or has no allowed_transitions
    """
    if policy is None:
        raise ValueError("Cannot generate reminder: policy is None")
    
    if not policy.allowed_transitions:
        raise ValueError("Cannot generate reminder: no allowed_transitions in policy")
    
    lines = [
        "",
        "---",
        "REMINDER: Emit exactly one of these tags (target names are literal, not placeholders):",
        ""
    ]
    
    for i, allowed in enumerate(policy.allowed_transitions, 1):
        tag = allowed.get("tag", "")
        target = allowed.get("target")
        
        # Build attributes list (excluding tag and target which are not attributes)
        attrs = []
        for key, value in allowed.items():
            if key not in ("tag", "target"):
                # Use appropriate quotes to avoid XML parsing issues
                # The parser supports both single and double quotes
                value_str = str(value)
                if '"' in value_str:
                    # Value contains double quote - use single quotes
                    # Note: If value contains both quote types, this will still work
                    # as the parser stops at the first quote, but this edge case
                    # shouldn't occur in practice (filenames don't contain quotes)
                    attrs.append(f"{key}='{value_str}'")
                else:
                    # No double quotes - use double quotes (standard)
                    # Single quotes in value are fine with double-quoted attribute
                    attrs.append(f'{key}="{value_str}"')
        
        # Build the complete XML tag string
        if tag == "result":
            # Result tag: <result>...</result> with variable payload
            tag_str = f"<{tag}>...</{tag}>"
            description = ""
        else:
            # Other tags: need target, may have attributes
            if not target:
                # Missing target is unusual but handle gracefully
                target = "TARGET"
            
            if attrs:
                # Tag with attributes: <tag attr="value">target</tag>
                attrs_str = " ".join(attrs)
                tag_str = f"<{tag} {attrs_str}>{target}</{tag}>"
            else:
                # Tag without attributes: <tag>target</tag>
                tag_str = f"<{tag}>{target}</{tag}>"
            description = ""
        
        # Format the line
        if description:
            lines.append(f"{i}. {tag_str} {description}")
        else:
            lines.append(f"{i}. {tag_str}")
    
    lines.append("")
    lines.append("Emit exactly one of the above tags.")
    lines.append("---")
    
    return "\n".join(lines)

File: internal/policy/policy.go
package policy

import (
	"fmt"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/vector76/raymond/internal/parsing"
)

// stateExtensions lists the file extensions recognized as valid state targets.
var stateExtensions = map[string]bool{
	".md":  true,
	".sh":  true,
	".bat": true,
	".ps1": true,
}

// Frontmatter delimiter patterns anchored to the start of content.
// Go RE2 supports (?s) to make . match newlines.
var (
	frontmatterRe      = regexp.MustCompile(`(?s)^---[ \t]*\n(.+?)\n---[ \t]*\n`)
	emptyFrontmatterRe = regexp.MustCompile(`^---[ \t]*\n---[ \t]*\n`)
)

// Policy represents a workflow state's transition policy parsed from YAML frontmatter.
type Policy struct {
	AllowedTransitions []map[string]string // each entry has at least "tag"
	Model              string              // empty if not specified
	Effort             string              // empty if not specified
}

// PolicyViolationError is returned when a transition violates the state's policy.
type PolicyViolationError struct {
	msg string
}

func (e *PolicyViolationError) Error() string { return e.msg }

// ParseFrontmatter extracts YAML frontmatter from markdown content.
//
// Returns (nil, content, nil) when no frontmatter is present.
// Returns (nil, body, nil) when frontmatter is empty or contains only null/empty YAML.
// Returns (*Policy, body, nil) on success.
// Returns (nil, "", error) when frontmatter contains invalid YAML.
func ParseFrontmatter(content string) (*Policy, string, error) {
	// Normalize Windows line endings so the regex works regardless of how the
	// file was checked out. Python's text-mode file I/O does this automatically;
	// Go's os.ReadFile reads raw bytes, so we do it explicitly here.
	content = strings.ReplaceAll(content, "\r\n", "\n")

	// Try non-empty frontmatter first.
	if m := frontmatterRe.FindStringSubmatchIndex(content); m != nil {
		yamlContent := content[m[2]:m[3]]
		body := content[m[1]:]

		if strings.TrimSpace(yamlContent) == "" {
			return nil, body, nil
		}
		p, err := parseYAML(yamlContent)
		if err != nil {
			return nil, "", err
		}
		return p, body, nil
	}

	// Try empty frontmatter (--- immediately followed by ---).
	if m := emptyFrontmatterRe.FindStringIndex(content); m != nil {
		return nil, content[m[1]:], nil
	}

	// No frontmatter.
	return nil, content, nil
}

// yamlFrontmatter is the intermediate struct for YAML unmarshaling.
type yamlFrontmatter struct {
	AllowedTransitions []map[string]string `yaml:"allowed_transitions"`
	Model              string              `yaml:"model"`
	Effort             string              `yaml:"effort"`
}

// parseYAML converts a YAML string into a Policy, or nil if the document is empty.
func parseYAML(yamlContent string) (*Policy, error) {
	// First check whether the document is null / empty (matches Python's `if not data`).
	var raw interface{}
	if err := yaml.Unmarshal([]byte(yamlContent), &raw); err != nil {
		return nil, fmt.Errorf("Invalid YAML frontmatter: %w", err)
	}
	if raw == nil {
		return nil, nil
	}
	if m, ok := raw.(map[string]interface{}); ok && len(m) == 0 {
		return nil, nil
	}

	var data yamlFrontmatter
	if err := yaml.Unmarshal([]byte(yamlContent), &data); err != nil {
		return nil, fmt.Errorf("Invalid YAML frontmatter: %w", err)
	}

	// Filter: only keep entries that have a "tag" key.
	var valid []map[string]string
	for _, entry := range data.AllowedTransitions {
		if _, ok := entry["tag"]; ok {
			valid = append(valid, entry)
		}
	}

	// Normalize model: lowercase + trim; empty string stays empty.
	model := strings.TrimSpace(strings.ToLower(data.Model))

	// Normalize effort: lowercase + trim.
	effort := strings.TrimSpace(strings.ToLower(data.Effort))

	return &Policy{
		AllowedTransitions: valid,
		Model:              model,
		Effort:             effort,
	}, nil
}

// TargetsMatch reports whether policyTarget (from the policy rule) matches
// transitionTarget (from the parsed transition).
//
// If policyTarget has a file extension, only an exact match is accepted.
// If policyTarget has no extension (abstract), it matches any transitionTarget
// that has the same stem and a recognized state extension (.md, .sh, .bat, .ps1).
func TargetsMatch(policyTarget, transitionTarget string) bool {
	if policyTarget == transitionTarget {
		return true
	}

	policyExt := filepath.Ext(policyTarget)
	if policyExt != "" {
		// Explicit extension — only exact match, which already failed above.
		return false
	}

	// Abstract target: check stem equality and valid extension.
	transitionExt := strings.ToLower(filepath.Ext(transitionTarget))
	transitionStem := strings.TrimSuffix(transitionTarget, filepath.Ext(transitionTarget))
	return transitionStem == policyTarget && stateExtensions[transitionExt]
}

// ShouldUseReminderPrompt reports whether a reminder prompt should be sent to
// the agent on parse failure. Reminders are only useful when the policy
// specifies allowed transitions; without them there is nothing to remind about.
func ShouldUseReminderPrompt(p *Policy) bool {
	return p != nil && len(p.AllowedTransitions) > 0
}

// ValidateTransitionPolicy checks that transition complies with p.
// A nil policy or a policy with an empty AllowedTransitions list imposes no
// restrictions. Returns a *PolicyViolationError on failure.
func ValidateTransitionPolicy(transition parsing.Transition, p *Policy) error {
	if p == nil || len(p.AllowedTransitions) == 0 {
		return nil
	}

	for _, allowed := range p.AllowedTransitions {
		if allowed["tag"] != transition.Tag {
			continue
		}
		// result tags: if the policy specifies a fixed payload, only match
		// when the transition's payload matches; otherwise allow any payload.
		if transition.Tag == "result" {
			if allowedPayload, ok := allowed["payload"]; ok {
				if strings.TrimSpace(transition.Payload) != allowedPayload {
					continue
				}
			}
			return nil
		}
		// Check target (if policy specifies one).
		if pt, ok := allowed["target"]; ok {
			if !TargetsMatch(pt, transition.Target) {
				continue
			}
		}
		// Check return attribute (call/function).
		if pr, ok := allowed["return"]; ok {
			if !TargetsMatch(pr, transition.Attributes["return"]) {
				continue
			}
		}
		// Check next attribute (fork, await).
		if pn, ok := allowed["next"]; ok {
			if !TargetsMatch(pn, transition.Attributes["next"]) {
				continue
			}
		}
		// Check timeout_next attribute (await).
		if ptn, ok := allowed["timeout_next"]; ok {
			if !TargetsMatch(ptn, transition.Attributes["timeout_next"]) {
				continue
			}
		}
		return nil // all checks passed
	}

	// No matching rule found — build a helpful error message.
	var allowedForTag []map[string]string
	for _, allowed := range p.AllowedTransitions {
		if allowed["tag"] == transition.Tag {
			allowedForTag = append(allowedForTag, allowed)
		}
	}
	if len(allowedForTag) > 0 {
		var detail string
		if transition.Tag == "result" {
			detail = fmt.Sprintf(
				"result with payload %q is not allowed. "+
					"Allowed payloads for %q: %v",
				transition.Payload, transition.Tag, allowedForTag,
			)
		} else {
			detail = fmt.Sprintf(
				"transition %q with target %q and attributes %v is not allowed. "+
					"Allowed combinations for %q: %v",
				transition.Tag, transition.Target, transition.Attributes,
				transition.Tag, allowedForTag,
			)
		}
		return &PolicyViolationError{msg: detail}
	}

	seen := map[string]bool{}
	var allowedTags []string
	for _, allowed := range p.AllowedTransitions {
		tag := allowed["tag"]
		if !seen[tag] {
			allowedTags = append(allowedTags, tag)
			seen[tag] = true
		}
	}
	return &PolicyViolationError{msg: fmt.Sprintf(
		"tag %q is not allowed. Allowed tags: %v",
		transition.Tag, allowedTags,
	)}
}

// CanUseImplicitTransition reports whether the policy's single allowed
// transition can be used implicitly (without requiring the agent to emit a tag).
//
// Conditions: policy exists, exactly one allowed transition, and either a
// target is specified (for non-result tags) or a fixed payload is specified
// (for result tags).
func CanUseImplicitTransition(p *Policy) bool {
	if p == nil || len(p.AllowedTransitions) != 1 {
		return false
	}
	allowed := p.AllowedTransitions[0]
	if allowed["tag"] == "result" {
		_, hasPayload := allowed["payload"]
		return hasPayload
	}
	// Await transitions cannot be implicit — the LLM must compose the
	// prompt inside the tag body.
	if allowed["tag"] == "await" {
		return false
	}
	_, hasTarget := allowed["target"]
	return hasTarget
}

// GetImplicitTransition constructs a Transition from the single allowed rule.
// Panics (via error return) when CanUseImplicitTransition would return false.
func GetImplicitTransition(p *Policy) (parsing.Transition, error) {
	if !CanUseImplicitTransition(p) {
		return parsing.Transition{}, fmt.Errorf(
			"cannot get implicit transition: policy must have exactly one " +
				"allowed transition with a target or a result with a fixed payload",
		)
	}
	allowed := p.AllowedTransitions[0]
	if allowed["tag"] == "result" {
		return parsing.Transition{
			Tag:     "result",
			Payload: allowed["payload"],
		}, nil
	}
	attrs := make(map[string]string)
	for k, v := range allowed {
		if k != "tag" && k != "target" {
			attrs[k] = v
		}
	}
	return parsing.Transition{
		Tag:        allowed["tag"],
		Target:     allowed["target"],
		Attributes: attrs,
		Payload:    "",
	}, nil
}

// GenerateReminderPrompt produces a formatted reminder message listing all
// allowed transitions. Used when an agent fails to emit a valid transition tag.
func GenerateReminderPrompt(p *Policy) (string, error) {
	if p == nil {
		return "", fmt.Errorf("cannot generate reminder: policy is nil")
	}
	if len(p.AllowedTransitions) == 0 {
		return "", fmt.Errorf("cannot generate reminder: no allowed_transitions in policy")
	}

	lines := []string{
		"",
		"---",
		"REMINDER: Emit exactly one of these tags (target names are literal, not placeholders):",
		"",
	}

	for i, allowed := range p.AllowedTransitions {
		tag := allowed["tag"]
		target := allowed["target"]

		// Collect and sort attribute keys (excluding tag and target) for deterministic output.
		var attrKeys []string
		for k := range allowed {
			if k != "tag" && k != "target" {
				attrKeys = append(attrKeys, k)
			}
		}
		sort.Strings(attrKeys)

		var attrParts []string
		for _, k := range attrKeys {
			v := allowed[k]
			if strings.Contains(v, `"`) {
				attrParts = append(attrParts, fmt.Sprintf("%s='%s'", k, v))
			} else {
				attrParts = append(attrParts, fmt.Sprintf(`%s="%s"`, k, v))
			}
		}

		var tagStr string
		if tag == "result" {
			if payload, ok := allowed["payload"]; ok {
				tagStr = fmt.Sprintf("<%s>%s</%s>", tag, payload, tag)
			} else {
				tagStr = fmt.Sprintf("<%s>...</%s>", tag, tag)
			}
		} else if tag == "await" {
			// Await: all non-tag keys become attributes (next, timeout,
			// timeout_next); the body is a prompt placeholder.  attrParts
			// already contains these since await entries have no "target" key.
			if len(attrParts) > 0 {
				tagStr = fmt.Sprintf("<await %s>[human-facing prompt here]</await>", strings.Join(attrParts, " "))
			} else {
				tagStr = "<await>[human-facing prompt here]</await>"
			}
		} else {
			if target == "" {
				target = "TARGET"
			}
			if len(attrParts) > 0 {
				tagStr = fmt.Sprintf("<%s %s>%s</%s>", tag, strings.Join(attrParts, " "), target, tag)
			} else {
				tagStr = fmt.Sprintf("<%s>%s</%s>", tag, target, tag)
			}
		}
		lines = append(lines, fmt.Sprintf("%d. %s", i+1, tagStr))
	}

	lines = append(lines, "", "Emit exactly one of the above tags.", "---")
	return strings.Join(lines, "\n"), nil
}
